// @cloudbase/functions-framework/build/async-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.asyncContext = void 0;
exports.newExecutionContext = newExecutionContext;
exports.wrapWithAsyncContext = wrapWithAsyncContext;
exports.getCurrentAsyncContext = getCurrentAsyncContext;
exports.isInFrameworkAsyncContext = isInFrameworkAsyncContext;
exports.getCurrentEventID = getCurrentEventID;
exports.done = done;
const node_async_hooks_1 = require("node:async_hooks");
const CONTEXT_ID = Symbol('@cloudbase/functions-framework');
exports.asyncContext = new node_async_hooks_1.AsyncLocalStorage();
function newExecutionContext(eventID) {
    return {
        id: CONTEXT_ID,
        done: false,
        eventID: eventID
    };
}
async function wrapWithAsyncContext(eventID, fn) {
    return exports.asyncContext.run(newExecutionContext(eventID), fn);
}
function getCurrentAsyncContext() {
    return exports.asyncContext.getStore();
}
function isInFrameworkAsyncContext() {
    // 注意：异步上下文嵌套将会出现覆盖，用户函数中使用 async_hooks 将会有副作用，可以禁用 async_hooks 解决
    // asyncLocalStorage.run({ id: 'a' }, () => {
    //   console.log(asyncLocalStorage.getStore()) // { id: 'a' }
    //   asyncLocalStorage.run({ id: 'b' }, () => {
    //     console.log(asyncLocalStorage.getStore()) // { id: 'b' }
    //   })
    // })
    return exports.asyncContext.getStore() !== undefined
        && exports.asyncContext.getStore()?.id === CONTEXT_ID;
}
function getCurrentEventID() {
    const currentContext = exports.asyncContext.getStore();
    return currentContext?.eventID || '';
}
async function done() {
    const currentContext = exports.asyncContext.getStore();
    if (currentContext !== undefined) {
        currentContext.done = true;
    }
}
