// @cloudbase/functions-framework/build/cli.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runCLI = runCLI;
const dotenvx = __importStar(require("@dotenvx/dotenvx"));
const fs_1 = require("fs");
const fs = __importStar(require("fs"));
const path_1 = require("path");
const path = __importStar(require("path"));
const framework_1 = require("./framework");
const functions_config_1 = require("./functions-config");
const logger_1 = require("./logger");
const options_1 = require("./options");
const router_1 = require("./router");
const tools_1 = require("./tools");
async function runCLI() {
    let options;
    try {
        options = (0, options_1.parseOptions)();
    }
    catch (err) {
        logger_1.frameworkLogger.error(`Failed to parse options: ${err}`);
        (0, tools_1.killInstance)();
        return;
    }
    if (options.printHelp) {
        logger_1.frameworkLogger.info(options_1.helpText);
        process.forceExit(0);
    }
    // 支持加载 .env 文件，需要主动开启
    // 配置类功能，主要用于测试场景，正式环境通常通过平台环境变量方式配置
    // .env 文件各运行环境通常会不一样
    if (options.dotEnvFilePath !== '') {
        dotenvx.config({
            path: options.dotEnvFilePath.split(','),
            override: true,
            strict: false,
            debug: false,
            logLevel: 'info'
        });
    }
    // 一次拉起所有函数时，存储支持多函数的 Router 实例
    const routers = new Map();
    async function handleOneFunction(fnName, options, functionsConfigFile) {
        let multiFnConfig;
        if ((0, fs_1.existsSync)(functionsConfigFile)) {
            try {
                multiFnConfig = (0, functions_config_1.loadFunctionsConfig)(functionsConfigFile);
                // 根据配置文件的路径，计算出函数根目录的绝对路径，以支持在任意目录执行命令
                multiFnConfig.functionsRoot = (0, path_1.join)((0, path_1.dirname)(functionsConfigFile), multiFnConfig.functionsRoot);
                logger_1.frameworkLogger.info(`Load user function '${fnName}' FunctionsConfig(${functionsConfigFile}):`, JSON.stringify(multiFnConfig));
            }
            catch (err) {
                logger_1.frameworkLogger.error(`Failed to load functions config: ${err}`);
                (0, tools_1.killInstance)();
                return;
            }
        }
        if (multiFnConfig) {
            routers.set(fnName, new router_1.Router(multiFnConfig));
        }
        const loadStartAt = Date.now();
        await (0, framework_1.loadAndRegisterFunctions)(fnName, options, multiFnConfig);
        const loadEndAt = Date.now();
        logger_1.frameworkLogger.info(`Load user function '${fnName}' cost: ${loadEndAt - loadStartAt}ms`);
    }
    if (options.loadAllFunctions && typeof options.functionsRoot === 'string' && options.functionsRoot !== '') {
        // 一次加载 options.functionsRoot 目录下的所有函数
        // 此时，--source 参数无效，实际 source 将基于 functionsRoot 目录产生
        const functionsRoot = (0, path_1.resolve)(options.functionsRoot);
        if (!fs.statSync(functionsRoot).isDirectory()) {
            throw new Error(`functionsRoot '${functionsRoot}' is not a directory`);
        }
        // 遍历 functionsRoot 目录下的所有函数目录
        for (const fnName of fs.readdirSync(functionsRoot)) {
            // 函数名不允许为 logs，因为 logs 是日志文件夹的名称
            if (fnName === 'logs') {
                continue;
            }
            const fnDir = path.join(functionsRoot, fnName);
            if (fs.statSync(fnDir).isDirectory()) {
                // 函数目录下的 cloudbase-functions.json 文件路径
                const functionsConfigFile = path.join(fnDir, options.functionsConfigFile);
                await handleOneFunction(fnName, { ...options, sourceLocation: fnDir }, functionsConfigFile);
            }
            else {
                logger_1.frameworkLogger.warn(`${fnName} not a directory, skip`);
            }
        }
    }
    else {
        if (options.functionsConfigFile && options.functionsConfigFile !== '') {
            if (options.sourceLocation !== '') {
                // 如果传了 sourceLocation 参数，则将 functionsConfigFile 转换为相对于 sourceLocation 的绝对路径
                // --source=examples/module-cjs 指定目录时，支持 functionsConfigFile 文件，因为可以找到文件
                // --source=examples/module-cjs/index.cjs 指定具体文件时，不支持 functionsConfigFile 文件，因为找不到文件
                options.functionsConfigFile = (0, path_1.resolve)(options.sourceLocation, options.functionsConfigFile);
            }
            else {
                // 如果没有传 sourceLocation 参数，则将 functionsConfigFile 转换为相对于 process.cwd() 的绝对路径
                options.functionsConfigFile = (0, path_1.resolve)(options.functionsConfigFile);
            }
        }
        // 只加载一个函数，fnName 为空字符串，不需要区分函数名，无法创建一个名称为 '' 的函数，比较安全
        await handleOneFunction('', { ...options }, options.functionsConfigFile);
    }
    await (0, framework_1.startFrameworkService)(options, routers);
}
