// @cloudbase/functions-framework/build/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorExt = void 0;
exports.decorateErrorStack = decorateErrorStack;
exports.newSysErr = newSysErr;
exports.newServerTimeoutErr = newServerTimeoutErr;
exports.newBadRequestErr = newBadRequestErr;
exports.newTooManyRequestsErr = newTooManyRequestsErr;
const constants_1 = require("./constants");
/**
 * 修饰错误堆栈
 * @param err 需要修饰错误实例
 * @returns
 */
function decorateErrorStack(err) {
    // 注意：Error 存在 Error、SyntaxError 或 自定义 Error 等多种类型，无法枚举
    if (!err || typeof err.message !== 'string' || typeof err.stack !== 'string') {
        return '';
    }
    const excludes = [constants_1.INNER_STACK_PREFIX, 'node_modules/pm2', 'node:'];
    return err.stack
        .split('\n')
        .filter(line => !excludes.some(exclude => line.includes(exclude)))
        .join('\n');
}
class ErrorExt extends Error {
    code = '';
    cause; // 因为 catch 到的可能不是 Error 这里可能不是 Error
    constructor(code, message, options) {
        super(message, options);
        this.code = code;
        this.cause = options.cause;
    }
}
exports.ErrorExt = ErrorExt;
function newSysErr(err) {
    return new ErrorExt('SYS_ERR', 'System Error.', { cause: err });
}
function newServerTimeoutErr(err) {
    return new ErrorExt('SERVER_TIMEOUT', 'Server Timeout.', { cause: err });
}
function newBadRequestErr(err) {
    return new ErrorExt('BAD_REQUEST', 'Bad Request.', { cause: err });
}
function newTooManyRequestsErr(err) {
    return new ErrorExt('TOO_MANY_REQUESTS', 'Too Many Requests.', { cause: err });
}
