// @cloudbase/functions-framework/build/framework.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startFrameworkService = startFrameworkService;
exports.loadAndRegisterFunctions = loadAndRegisterFunctions;
exports.loadFunctions = loadFunctions;
const function_loader_1 = require("./function-loader");
const function_registry_1 = require("./function-registry");
const logger_1 = require("./logger");
const server_1 = require("./server");
const tools_1 = require("./tools");
async function startFrameworkService(options, routes) {
    Error.stackTraceLimit = 100;
    if (options.interceptOutput) {
        logger_1.Interceptor.intercept();
    }
    if (options.noExit) {
        (0, tools_1.interceptProcessExit)();
    }
    if (options.dryRun) {
        process.forceExit(0);
    }
    const opt = {
        dirname: options.logDirname,
        maxSize: options.maxLogFileSize,
        maxFiles: options.maxLogFiles
    };
    logger_1.functionsLogger.init([
        { type: logger_1.FunctionsLoggerType.ACCESS, options: opt },
        { type: logger_1.FunctionsLoggerType.USERCODE, options: opt }
    ]);
    if (options.logContentEnabled) {
        (0, logger_1.setLogContentEnabled)(true);
    }
    const server = (0, server_1.createServer)(options, routes);
    const signalHandler = new tools_1.ServerSignalHandler(server, options.gracefulShutdown);
    server.listen(options.port, () => {
        signalHandler.setup();
        logger_1.frameworkLogger.info(`Server listening on port ${options.port}, Visit: http://127.0.0.1:${options.port}/`);
        logger_1.frameworkLogger.info(`Framework options: ${JSON.stringify(options)}`);
    });
    server.timeout = 0;
    server.keepAliveTimeout = options.keepAliveTimeoutMS;
}
/**
 * 加载并注册函数
 */
async function loadAndRegisterFunctions(fnName, options, multiFnConfig) {
    const loadResults = await loadFunctions(options, multiFnConfig);
    if (Array.isArray(loadResults)) {
        for (const loadResult of loadResults) {
            if (!loadResult?.userFunction) {
                logger_1.frameworkLogger.error(`Failed to load function ${loadResult?.name}, because "${loadResult?.reason}"`);
                process.forceExit(1);
            }
            (0, function_registry_1.event)((0, function_registry_1.buildFnId)(fnName || '', loadResult.name), loadResult.userFunction, loadResult.pkgInfo);
        }
        logger_1.frameworkLogger.info(`Load user function '${fnName}' child functions cost: ${loadResults.map(loadResult => `${(0, function_registry_1.buildFnId)(fnName || '', loadResult.name)}(${loadResult.timeCost}ms)`).join(', ')}`);
    }
    else {
        if (!loadResults?.userFunction) {
            logger_1.frameworkLogger.error(`Failed to load target function, because "${loadResults?.reason}"`);
            process.forceExit(1);
        }
        (0, function_registry_1.event)((0, function_registry_1.buildFnId)(fnName || '', loadResults.name), loadResults.userFunction, loadResults.pkgInfo);
    }
}
/**
 * 加载单个或多个函数
 * @param options FrameworkOptions
 * @returns 加载结果
 */
async function loadFunctions(options, multiFnConfig) {
    let nameOrFunctions;
    let codeLocation;
    if (multiFnConfig) {
        const { functionsRoot, functions } = multiFnConfig;
        nameOrFunctions = functions;
        codeLocation = functionsRoot;
    }
    else {
        nameOrFunctions = function_loader_1.DEFAULT_FN_NAME;
        codeLocation = options.sourceLocation;
    }
    return await (0, function_loader_1.loadUserFunctions)(nameOrFunctions, codeLocation, options.target);
}
