// @cloudbase/functions-framework/build/function-wrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapHandlerFunction = wrapHandlerFunction;
const error_1 = require("./error");
const function_sse_1 = require("./function-sse");
const unified_responder_1 = require("./unified-responder");
const utils_lang_1 = require("./utils.lang");
const wrapEventFunction = (execute, pkgInfo, captureStack) => {
    return async (ctx) => {
        let event;
        if (ctx.request.files) {
            // form-data, restructured to form-data like object
            event = { ...ctx.request.files, ...ctx.request.body };
        }
        else if (ctx.request.body) {
            // normal json, text etc. that can be parsed
            event = ctx.request.body;
        }
        else {
            // unrecognized, send raw body
            event = ctx.rawBody;
        }
        let executeDone = false;
        let hasSwitchSSEMode = false;
        const context = {
            ctxId: ctx.state.eventID,
            eventID: ctx.state.eventID,
            eventType: ctx.state.eventType,
            timestamp: ctx.state.timestamp,
            httpContext: ctx.httpContext,
            sse: (options = {}) => {
                // 如果已经切换成了 sse，直接返回同一个实例即可
                const sseInstance = ctx.sse?.(options);
                if (sseInstance) {
                    return sseInstance;
                }
                // 已经响应或函数执行完，无法切换到SSE
                if (ctx.res.headersSent || executeDone) {
                    return null;
                }
                hasSwitchSSEMode = true;
                const sse = new function_sse_1.ServerSentEvent(ctx, options);
                // 保存对当前对象的引用，保证后续同 ctx 取到的 sse 对象是同一个
                ctx.sse = () => sse;
                return sse;
            }
        };
        if (Object.keys(ctx.extendedContext || {}).length > 0) {
            context.extendedContext = ctx.extendedContext;
        }
        (0, utils_lang_1.deepFreeze)(context);
        ctx.state.event = event;
        ctx.state.contextInjected = context;
        ctx.state.codeInfo = {
            name: pkgInfo?.name || '',
            version: pkgInfo?.version || ''
        };
        let result;
        const start = Date.now();
        try {
            result = await execute(event, context);
        }
        catch (e) {
            ctx.state.error = e;
            // 单独处理 SSE 情况，避免客户端 hang 住
            if (hasSwitchSSEMode) {
                const sse = ctx.sse?.();
                if (sse && !sse.closed) {
                    // 出错时，发送 error 事件并关闭连接
                    // https://html.spec.whatwg.org/multipage/server-sent-events.html#sse-processing-model
                    const errorEvent = {
                        event: 'error',
                        data: e.message,
                        comment: 'uncaught-exception-from-user-code'
                    };
                    sse.emit('error', errorEvent);
                    sse.end(errorEvent);
                }
            }
        }
        finally {
            ctx.state.userCodeTimeCost = Date.now() - start;
        }
        executeDone = true;
        if (hasSwitchSSEMode) {
            return;
        }
        const e = ctx.state.error ? (0, error_1.newSysErr)(ctx.state.error) : null;
        return (0, unified_responder_1.sendResponse)(ctx, result, e, undefined, captureStack);
    };
};
function wrapHandlerFunction(userFunction, pkgInfo, type, captureStack = false) {
    switch (type) {
        case 'event':
            return wrapEventFunction(userFunction, pkgInfo, captureStack);
        default:
            throw new Error(`Unsupported function type: ${type}`);
    }
}
