// @cloudbase/functions-framework/build/function-registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.event = void 0;
exports.register = register;
exports.getRegisteredFunction = getRegisteredFunction;
exports.getAllRegisteredFunctions = getAllRegisteredFunctions;
exports.buildFnId = buildFnId;
const registrationContainer = new Map();
function register(name, signatureType, userFunction, pkgInfo) {
    registrationContainer.set(name, { name, signatureType, userFunction, pkgInfo });
}
function getRegisteredFunction(name) {
    return registrationContainer.get(name);
}
function getAllRegisteredFunctions() {
    return Array.from(registrationContainer.values());
}
/**
 * 注册一个事件函数
 * @param functionName 函数名称
 * @param handler 事件处理函数
 */
const event = (fnId, handler, pkgInfo) => {
    if (!isFullFnId(fnId)) {
        fnId = buildFnId('', fnId);
    }
    register(fnId, 'event', handler, pkgInfo);
};
exports.event = event;
function isFullFnId(fnId) {
    return fnId.split('.').length === 2;
}
function buildFnId(fnName, childFnName) {
    return `${fnName}.${childFnName}`;
}
