// @cloudbase/functions-framework/build/function-loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadUserFunctionForTesting = exports.DEFAULT_FN_NAME = void 0;
exports.loadUserFunctions = loadUserFunctions;
const path_1 = __importDefault(require("path"));
const read_pkg_up_1 = __importDefault(require("read-pkg-up"));
const url_1 = require("url");
const function_registry_1 = require("./function-registry");
const logger_1 = require("./logger");
exports.DEFAULT_FN_NAME = ''; // 用户不可能定义一个空函数名
function readPkgInfo(modulePath) {
    return (0, read_pkg_up_1.default)({
        cwd: path_1.default.dirname(modulePath),
        normalize: false
    });
}
async function isEsModule(modulePath) {
    const ext = path_1.default.extname(modulePath);
    if (ext === '.mjs') {
        return true;
    }
    if (ext === '.cjs') {
        return false;
    }
    const pkg = await (0, read_pkg_up_1.default)({
        cwd: path_1.default.dirname(modulePath),
        normalize: false
    });
    // If package.json specifies type as 'module', it's an ES module.
    return pkg?.packageJson.type === 'module';
}
const dynamicImport = new Function('modulePath', 'return import(modulePath)'
// eslint-disable-next-line @typescript-eslint/no-explicit-any
);
/**
 * 加载用户函数
 * @param nameOrFunctions 单个函数名称或函数定义列表
 * @param codeLocation 函数文件
 * @param functionTarget 目标函数名称
 * @returns
 */
async function loadUserFunctions(nameOrFunctions, codeLocation, functionTarget) {
    if (Array.isArray(nameOrFunctions)) {
        const functionPromises = nameOrFunctions.map(func => loadOneFunction(func.name, func.source ? path_1.default.resolve(codeLocation, func.directory, func.source) : path_1.default.resolve(codeLocation, func.directory), func.target || functionTarget));
        return Promise.all(functionPromises);
    }
    else {
        return loadOneFunction(nameOrFunctions, codeLocation, functionTarget);
    }
}
/**
 * 加载单个函数
 * @param name 函数名称
 * @param codeLocation 函数文件
 * @param functionTarget 目标函数名称
 * @returns
 */
async function loadOneFunction(name, codeLocation, functionTarget) {
    const start = Date.now();
    try {
        const functionModulePath = getFunctionModulePath(codeLocation);
        if (functionModulePath === null) {
            return {
                timeCost: Date.now() - start,
                name,
                pkgInfo: {},
                reason: `Provided code location '${codeLocation}' is not a loadable module.`
            };
        }
        let functionModule;
        const esModule = await isEsModule(functionModulePath);
        if (esModule) {
            // Resolve module path to file:// URL. Required for windows support.
            const fpath = (0, url_1.pathToFileURL)(functionModulePath);
            functionModule = await dynamicImport(fpath.href);
        }
        else {
            functionModule = require(functionModulePath);
        }
        // If the customer declaratively registered a function matching the target
        // return that.
        const registeredFunction = (0, function_registry_1.getRegisteredFunction)(functionTarget);
        if (registeredFunction) {
            return {
                ...registeredFunction,
                timeCost: Date.now() - start
            };
        }
        const userFunction = functionTarget
            .split('.')
            .reduce((code, functionTargetPart) => {
            if (typeof code === 'undefined') {
                return undefined;
            }
            else {
                return code[functionTargetPart];
            }
        }, functionModule);
        if (typeof userFunction === 'undefined') {
            return {
                name,
                timeCost: Date.now() - start,
                pkgInfo: {},
                reason: `Function '${functionTarget}' is not defined in the provided module.`
            };
        }
        if (typeof userFunction !== 'function') {
            return {
                name,
                pkgInfo: {},
                timeCost: Date.now() - start,
                reason: `'${functionTarget}' needs to be of type function. Got: ${typeof userFunction}`
            };
        }
        let pkgInfo;
        try {
            pkgInfo = await readPkgInfo(functionModulePath);
        }
        catch (ex) {
            logger_1.frameworkLogger.error('Failed to read package.json', ex);
        }
        return {
            name,
            pkgInfo: pkgInfo?.packageJson || {},
            userFunction: userFunction,
            timeCost: Date.now() - start
        };
    }
    catch (ex) {
        const err = ex;
        let additionalHint;
        if (isNodeError(err) && err.code === 'MODULE_NOT_FOUND') {
            additionalHint =
                'Did you list all required modules in the package.json dependencies?\n';
        }
        else {
            additionalHint = 'Is there a syntax error in your code?\n';
        }
        return {
            name,
            pkgInfo: {},
            reason: `Provided module can't be loaded.\n${additionalHint}` +
                `Detailed stack trace: ${err.stack}`,
            timeCost: Date.now() - start
        };
    }
}
/**
* Returns resolved path to the module containing the user function.
* Returns null if the module can not be identified.
* @param codeLocation - Directory with user's code
* @return Resolved path or null.
*/
function getFunctionModulePath(codeLocation) {
    try {
        return require.resolve(codeLocation);
    }
    catch (ex) {
        // Ignore exception, this means the function was not found here.
        // textLogger.error('Failed to resolve code location.', ex)
    }
    return null;
}
const isNodeError = (error) => error instanceof Error;
/**
 * This should only be accessed in test files
 */
exports.loadUserFunctionForTesting = loadOneFunction;
