// @cloudbase/functions-framework/build/function-sse.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerSentEvent = void 0;
const events_1 = __importDefault(require("events"));
const stream_1 = require("stream");
const util = __importStar(require("util"));
const utils_lang_1 = require("./utils.lang");
const kDefaultEncoder = new util.TextEncoder();
const KEEPALIVE_INTERVAL_MS = 3000;
const HEARTBEAT_MESSAGE = Buffer.from(': heartbeat\n\n');
class ServerSentEvent extends events_1.default {
    ctx;
    stream;
    closed = false;
    encoder = kDefaultEncoder;
    keepaliveTimer;
    constructor(ctx, options = {}) {
        super();
        this.ctx = ctx;
        this.stream = new stream_1.PassThrough({
            objectMode: false
        });
        this.stream.pipe(ctx.res);
        ctx.respond = false;
        ctx.req.once('close', () => {
            this.emit('close');
            this.closed = true;
        });
        ctx.req.socket.setTimeout(0);
        ctx.req.socket.setNoDelay(true);
        ctx.req.socket.setKeepAlive(true);
        // 支持自定义响应头
        if (options && options.headers) {
            ctx.set(options.headers);
        }
        ctx.set({
            'Content-Type': 'text/event-stream; charset=utf-8',
            'Cache-Control': 'no-cache',
            Connection: 'keep-alive',
            'X-Request-Id': ctx.state.eventID
        });
        ctx.status = 200;
        ctx.flushHeaders();
        // 隐藏属性
        for (const key of ['ctx', 'stream', 'encoder']) {
            Object.defineProperty(this, key, {
                configurable: false,
                enumerable: false,
                writable: false
            });
        }
        // 默认开启心跳
        if (!Reflect.has(options, 'keepalive')) {
            options.keepalive = true;
        }
        if (options.keepalive) {
            this.keepalive();
        }
    }
    /**
     * 启动心跳定时器
     */
    keepalive() {
        if (this.keepaliveTimer) {
            clearInterval(this.keepaliveTimer);
        }
        this.keepaliveTimer = setInterval(() => {
            if (this.closed) {
                return;
            }
            this.send(HEARTBEAT_MESSAGE);
        }, KEEPALIVE_INTERVAL_MS);
    }
    setEncoder(encoder) {
        this.encoder = encoder;
    }
    send(event) {
        const encoder = this.encoder;
        const stream = this.stream;
        if (this.closed) {
            return false;
        }
        const events = Array.isArray(event) ? event : [event];
        const bufs = [];
        for (const event of events) {
            if (Buffer.isBuffer(event)) {
                bufs.push(event);
            }
            else if (typeof event === 'string') {
                bufs.push(Buffer.from(encoder.encode(event)));
            }
            else if ((0, utils_lang_1.isPlainObject)(event)) {
                const msgs = eventToMsgs(event);
                for (const msg of msgs) {
                    if (Buffer.isBuffer(msg)) {
                        bufs.push(msg);
                    }
                    else {
                        bufs.push(Buffer.from(encoder.encode(msg)));
                    }
                }
            }
        }
        const buf = Buffer.concat(bufs);
        if (Buffer.byteLength(buf) > 0) {
            return stream.write(Buffer.concat(bufs));
        }
        return false;
    }
    end(msg) {
        this.send(msg);
        this.stream.end();
        this.ctx.res.end();
        this.closed = true;
        if (this.keepaliveTimer) {
            clearInterval(this.keepaliveTimer);
            this.keepaliveTimer = undefined;
        }
    }
}
exports.ServerSentEvent = ServerSentEvent;
function eventToMsgs(event) {
    const msgs = [];
    if ('comment' in event) {
        msgs.push(`: ${event.comment}\n`);
    }
    if ('event' in event) {
        msgs.push(`event: ${event.event}\n`);
    }
    if ('id' in event) {
        msgs.push(`id: ${event.id}\n`);
    }
    if ('retry' in event) {
        msgs.push(`retry: ${event.retry}\n`);
    }
    if ('data' in event) {
        msgs.push(toDataEventString(event.data));
    }
    // 消息之间多加一个换行符号
    if (msgs.length > 0) {
        msgs.push('\n');
    }
    return msgs;
}
function toDataEventString(data) {
    if (Buffer.isBuffer(data)) {
        return Buffer.concat([Buffer.from('data: '), data, Buffer.from('\n')]);
    }
    if ((0, utils_lang_1.isPlainObject)(data)) {
        return toDataEventString(JSON.stringify(data));
    }
    const dataStr = typeof data !== 'string' ? `${data}` : data;
    // https://html.spec.whatwg.org/multipage/server-sent-events.html
    // Example:
    //
    // data: This is the first message.
    //
    // data: This is the second message, it
    // data: has two lines.
    //
    // data: This is the third message.
    // 规范化换行符，同一个事件中的换行符只保留一个 '\n'
    // 另外，不要 trim 掉前后空格，这是消息内容的一部分
    // Q：如果换行符是消息内容的一部分这样处理会不会有问题？
    // A：因为换行符在 SSE 消息中有特殊含义，所以消息中不能包含换行符，如果消息中存在换行符号，需转义
    // data = 'aaa\r\r\r\rbbb\n\r\n\nccc\n\r   \r\n  \n\n\n  dddd '
    // 'data: aaa\ndata: bbb\ndata: ccc\ndata:    \ndata:   \ndata:   dddd \n'
    return dataStr
        .split(/[\r\n]+/)
        .map((line) => `data: ${line}\n`)
        .join('');
}
