// @cloudbase/functions-framework/build/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.functionsLogger = exports.FunctionsLogger = exports.FunctionsLoggerType = exports.frameworkLogger = exports.FrameworkLogger = exports.Interceptor = void 0;
exports.setLogContentEnabled = setLogContentEnabled;
const console_1 = require("console");
const path_1 = __importDefault(require("path"));
const stream_1 = require("stream");
const util_1 = __importDefault(require("util"));
const winston_1 = require("winston");
const winston_daily_rotate_file_1 = __importDefault(require("winston-daily-rotate-file"));
const async_context_1 = require("./async-context");
const constants_1 = require("./constants");
const constants_2 = require("./constants");
const tools_1 = require("./tools");
const types_1 = require("./types");
const user_code_log_1 = require("./user-code-log");
class EmptyWritableStream extends stream_1.Writable {
    _write(_chunk, _encoding, callback) {
        callback();
    }
}
const emptyWritableStream = new EmptyWritableStream();
function truncateContent(content) {
    if (content.length > constants_2.USER_CODE_LOGS_MAX_CONTENT_LENGTH) {
        return content.slice(0, constants_2.USER_CODE_LOGS_MAX_CONTENT_LENGTH)
            + `... ${content.length - constants_2.USER_CODE_LOGS_MAX_CONTENT_LENGTH} more characters`;
    }
    return content;
}
var logType;
(function (logType) {
    logType["JSON"] = "json";
    logType["TEXT"] = "text";
})(logType || (logType = {}));
var logLevel;
(function (logLevel) {
    logLevel["INFO"] = "log";
    logLevel["WARN"] = "warn";
    logLevel["ERROR"] = "error";
})(logLevel || (logLevel = {}));
const consoleMethods = ['debug', 'info', 'log', 'trace', 'warn', 'error'];
const originalProcessStdoutWrite = process.stdout.write.bind(process.stdout);
const originalProcessStderrWrite = process.stderr.write.bind(process.stderr);
const originalConsole = {
    [logLevel.INFO]: console.info.bind(console),
    [logLevel.WARN]: console.warn.bind(console),
    [logLevel.ERROR]: console.error.bind(console)
};
class Interceptor {
    constructor() { }
    /**
     * 拦截标准输出，事件上下文格式化
     */
    static intercept() {
        Interceptor.interceptConsole();
        Interceptor.interceptOriginalProcessStdout();
        Interceptor.interceptOriginalProcessStderr();
    }
    static interceptConsole() {
        function callAddUserCodeLog(level, message, ...optionalParams) {
            if ((0, async_context_1.isInFrameworkAsyncContext)()) {
                const log = {
                    '@timestamp': new Date().toISOString(),
                    // 注意：重构代码注意 caller 变化
                    caller: (0, tools_1.getCaller)(callAddUserCodeLog),
                    level: level,
                    content: truncateContent(util_1.default.formatWithOptions({ colors: false }, message, ...optionalParams))
                };
                (0, user_code_log_1.addUserCodeLog)(log);
            }
        }
        // 劫持 console 输出
        const anotherConsole = new console_1.Console({
            stdout: emptyWritableStream,
            stderr: emptyWritableStream,
            colorMode: false,
            groupIndentation: 2,
            inspectOptions: {}
        });
        const originalConsole = global.console;
        for (const method of consoleMethods) {
            anotherConsole[method] = (message, ...optionalParams) => {
                if ((0, async_context_1.isInFrameworkAsyncContext)()) {
                    // 仅在 Framework 的异步上下文（主要是 main 函数中），劫持
                    callAddUserCodeLog(method, message, ...optionalParams);
                }
                else {
                    originalConsole[method](message, ...optionalParams);
                }
            };
        }
        // Q: other methods like console.dir how to handle? Keep the original behavior
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        global.originalConsole = global.console;
        global.console = anotherConsole;
    }
    static interceptOriginalProcessStdout() {
        process.stdout.write = function (...args) {
            if ((0, async_context_1.isInFrameworkAsyncContext)()) {
                return true;
            }
            else {
                return originalProcessStdoutWrite(...args);
            }
        };
    }
    static interceptOriginalProcessStderr() {
        process.stderr.write = function (...args) {
            if ((0, async_context_1.isInFrameworkAsyncContext)()) {
                return true;
            }
            else {
                return originalProcessStderrWrite(...args);
            }
        };
    }
}
exports.Interceptor = Interceptor;
class FrameworkLogger {
    prefix = constants_1.LOG_PREFIX;
    format;
    constructor({ format = logType.TEXT }) {
        this.format = format;
    }
    info(message, ...optionalParams) {
        this.log({
            format: this.format,
            level: logLevel.INFO,
            eventID: (0, async_context_1.getCurrentEventID)()
        }, message, ...optionalParams);
    }
    warn(message, ...optionalParams) {
        this.log({
            format: this.format,
            level: logLevel.WARN,
            eventID: (0, async_context_1.getCurrentEventID)()
        }, message, ...optionalParams);
    }
    error(message, ...optionalParams) {
        this.log({
            format: this.format,
            level: logLevel.ERROR,
            eventID: (0, async_context_1.getCurrentEventID)()
        }, message, ...optionalParams);
    }
    /**
     * 使用原始标准输出打印日志
     */
    static printStdout(message) {
        originalProcessStdoutWrite(message + '\n');
    }
    /**
     * 使用原始标准错误打印日志
     */
    static printStderr(message) {
        originalProcessStderrWrite(message + '\n');
    }
    formatPrefix(prefix, id) {
        return id ? `[${prefix}][pid:${process.pid}](${id})` : `[${prefix}][pid:${process.pid}]`;
    }
    log(options = { level: logLevel.INFO, format: logType.TEXT }, message, ...optionalParams) {
        if (options.eventID) {
            (0, user_code_log_1.addUserCodeLog)({ level: options.level, content: util_1.default.inspect(message) });
        }
        const method = options.level || logLevel.INFO;
        const prefix = this.formatPrefix(this.prefix, options.eventID);
        if (options.format === logType.JSON) {
            originalConsole[method](prefix, message, ...optionalParams);
        }
        else {
            originalConsole[method](prefix, message, ...optionalParams);
        }
    }
}
exports.FrameworkLogger = FrameworkLogger;
exports.frameworkLogger = new FrameworkLogger({ format: logType.TEXT });
var FunctionsLoggerType;
(function (FunctionsLoggerType) {
    FunctionsLoggerType["ACCESS"] = "accesslog";
    FunctionsLoggerType["USERCODE"] = "usercodelog";
})(FunctionsLoggerType || (exports.FunctionsLoggerType = FunctionsLoggerType = {}));
let logContentEnabled = false;
function setLogContentEnabled(enabled) {
    logContentEnabled = enabled;
}
class FunctionsLogger {
    tLoggers;
    constructor() {
        this.tLoggers = new Map();
    }
    init(opts) {
        opts.forEach(({ type, options }) => {
            this.tLoggers.set(type, createWinstonLogger({
                filename: type,
                dirname: options.dirname,
                maxSize: options.maxSize,
                maxFiles: options.maxFiles
            }));
        });
    }
    logAccesslog(level, message) {
        this.log(FunctionsLoggerType.ACCESS, level, message);
    }
    logUserCodelog() {
        // 注意：这里只能记录同步返回前的日志，同步返回后的日志这里服务记录
        const logs = (0, user_code_log_1.takeUserCodeLogs)();
        if (logs.length > 0) {
            const contents = [];
            for (const log of logs) {
                contents.push(log['content']);
            }
            const now = new Date().toISOString();
            const userCodeLog = {
                '@timestamp': now,
                'level': 'info',
                'module': '@cloudbase/functions-framework',
                'name': '',
                'logType': types_1.LogType.USERCODELOG,
                'eventId': (0, async_context_1.getCurrentEventID)(),
                'requestId': (0, async_context_1.getCurrentEventID)(),
                'scene': (0, async_context_1.getCurrentAsyncContext)()?.done ? 'logAfterReturn' : 'logBeforeReturn',
                'logs': logs
            };
            if (logContentEnabled) {
                userCodeLog['__CONTENT__'] = contents.join('\n');
            }
            this.log(FunctionsLoggerType.USERCODE, 'info', JSON.stringify(userCodeLog));
        }
    }
    log(type, level, message) {
        this.tLoggers.get(type)?.log(level, message);
    }
}
exports.FunctionsLogger = FunctionsLogger;
exports.functionsLogger = new FunctionsLogger();
function createWinstonLogger(options) {
    const dirname = options.dirname || constants_1.DEFAULT_LOG_DIRNAME;
    const transports = [
        new winston_daily_rotate_file_1.default({
            dirname: dirname,
            datePattern: 'YYYY-MM-DD-HH-mm',
            frequency: '720m', // unit of frequency should be coordinated with datePattern
            filename: `${options.filename}`,
            extension: '.log',
            auditFile: path_1.default.join(dirname, `${options.filename}.audit.json`),
            maxSize: options.maxSize || constants_1.DEFAULT_MAX_LOG_FILE_SIZE,
            maxFiles: options.maxFiles || constants_1.DEFAULT_MAX_LOG_FILES
        })
    ];
    const winstonLogger = (0, winston_1.createLogger)({
        exitOnError: false,
        level: 'info',
        transports: transports,
        format: winston_1.format.printf(({ message }) => message)
    });
    return winstonLogger;
}
