// @cloudbase/functions-framework/build/options.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseOptions = exports.helpText = exports.ConfigurableOption = void 0;
const minimist_1 = __importDefault(require("minimist"));
const path_1 = require("path");
const constants = __importStar(require("./constants"));
const logger_1 = require("./logger");
const types_1 = require("./types");
class ConfigurableOption {
    inputOption;
    alias;
    envVar;
    defaultValue;
    transform;
    constructor(inputOption, alias, envVar, defaultValue, transform) {
        this.inputOption = inputOption;
        this.alias = alias;
        this.envVar = envVar;
        this.defaultValue = defaultValue;
        this.transform = transform;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    parse(dict, envVars) {
        const value = dict[this.inputOption] || envVars[this.envVar] || this.defaultValue;
        return this.transform ? this.transform(value) : value;
    }
}
exports.ConfigurableOption = ConfigurableOption;
const PortOption = new ConfigurableOption(constants.OPTION_PORT, 'p', constants.ENV_PORT, constants.DEFAULT_PORT, (port) => {
    if (typeof port === 'string') {
        port = parseInt(port, 10);
    }
    if (isNaN(port) || port < 1 || port > 65535) {
        throw new Error(`Invalid port number ${port}.`);
    }
    return port;
});
const FunctionsRootOption = new ConfigurableOption(constants.OPTION_FUNCTIONS_ROOT, '', constants.ENV_FUNCTIONS_ROOT, constants.DEFAULT_FUNCTIONS_ROOT, (root) => {
    if (typeof root !== 'string') {
        return new Error(`Invalid functions root ${root}.`);
    }
    return root;
});
const enableLoadAllFunctionsOption = new ConfigurableOption(constants.OPTION_LOAD_ALL_FUNCTIONS, '', constants.ENV_LOAD_ALL_FUNCTIONS, constants.DEFAULT_LOAD_ALL_FUNCTIONS, (enableLoadAllFunctions) => {
    return enableLoadAllFunctions === 'true';
});
const FunctionTargetOption = new ConfigurableOption(constants.OPTION_TARGET, '', constants.ENV_TARGET, constants.DEFAULT_TARGET);
const FunctionSourceOption = new ConfigurableOption(constants.OPTION_SOURCE, '', constants.ENV_SOURCE, constants.DEFAULT_SOURCE, (source) => {
    if (typeof source !== 'string') {
        return new Error(`Invalid source location ${source}.`);
    }
    return (0, path_1.resolve)(source);
});
const SignatureOption = new ConfigurableOption(constants.OPTION_SIGNATURE_TYPE, 'type', constants.ENV_SIGNATURE_TYPE, constants.DEFAULT_SIGNATURE_TYPE, (type) => {
    if (!types_1.SignatureType.includes(type)) {
        return new Error(`Invalid signature type ${type}.`);
    }
    return type;
});
const TimeoutOption = new ConfigurableOption(constants.OPTION_TIMEOUT, '', constants.ENV_TIMEOUT, constants.DEFAULT_TIMEOUT, (timeout) => {
    if (typeof timeout === 'string') {
        timeout = parseInt(timeout, 10);
    }
    if (isNaN(timeout) || timeout < 0) {
        return new Error(`Invalid timeout value ${timeout}.`);
    }
    return timeout;
});
const GracefulShutDownOption = new ConfigurableOption(constants.OPTION_GRACEFUL_SHUTDOWN, '', constants.ENV_GRACEFUL_SHUTDOWN, process.env.NODE_ENV === 'production', (gracefulShutdown) => {
    if (typeof gracefulShutdown === 'string') {
        gracefulShutdown = gracefulShutdown.toLowerCase() === 'true';
    }
    return gracefulShutdown;
});
const KeepAliveTimeoutOption = new ConfigurableOption(constants.OPTION_KEEP_ALIVE_TIMEOUT, '', constants.ENV_KEEP_ALIVE_TIMEOUT, constants.DEFAULT_KEEP_ALIVE_TIMEOUT, (timeout) => {
    if (typeof timeout === 'string') {
        timeout = parseInt(timeout, 10);
    }
    if (isNaN(timeout) || timeout < 0) {
        return new Error(`Invalid keep alive timeout value ${timeout}.`);
    }
    return timeout;
});
const ConcurrencyLimitOption = new ConfigurableOption(constants.OPTION_CONCURRENCY_LIMIT, '', constants.ENV_CONCURRENCY_LIMIT, constants.DEFAULT_CONCURRENCY_LIMIT, (limit) => {
    if (typeof limit === 'string') {
        limit = parseInt(limit, 10);
    }
    return limit;
});
const NoExitOption = new ConfigurableOption(constants.OPTION_NO_EXIT, '', constants.ENV_NO_EXIT, constants.DEFAULT_NO_EXIT, (noExit) => {
    if (typeof noExit === 'string') {
        noExit = noExit.toLowerCase() === 'true';
    }
    return noExit;
});
const CaptureStackTrace = new ConfigurableOption(
// 如果将这里的 process.env.NODE_ENV 判断改为函数调用，测试会失败
constants.OPTION_CAPTURE_STACK, '', constants.ENV_CAPTURE_STACK, !(process.env.NODE_ENV === 'production'), (captureStack) => {
    if (typeof captureStack === 'string') {
        captureStack = captureStack.toLowerCase() === 'true';
    }
    return captureStack;
});
const LogHeaderBodyOption = new ConfigurableOption(constants.OPTION_LOG_HEADER_BODY, '', constants.ENV_LOG_HEADER_BODY, constants.DEFAULT_LOG_HEADER_BODY, (logHeaderBody) => {
    if (typeof logHeaderBody === 'string') {
        logHeaderBody = logHeaderBody.toLowerCase() === 'true';
    }
    return logHeaderBody;
});
const LogEventContextOption = new ConfigurableOption(constants.OPTIONS_LOG_EVENT_CONTEXT, '', constants.ENV_LOG_EVENT_CONTEXT, constants.DEFAULT_LOG_EVENT_CONTEXT, (logEventContext) => {
    if (typeof logEventContext === 'string') {
        logEventContext = logEventContext.toLowerCase() === 'true';
    }
    return logEventContext;
});
const LogContentEnabledOption = new ConfigurableOption(constants.OPTION_LOG_CONTENT_ENABLED, '', constants.ENV_LOG_CONTENT_ENABLED, constants.DEFAULT_LOG_CONTENT_ENABLED, (logContentEnabled) => {
    if (typeof logContentEnabled === 'string') {
        logContentEnabled = logContentEnabled.toLowerCase() === 'true';
    }
    return logContentEnabled;
});
const LogDirnameOption = new ConfigurableOption(constants.OPTIONS_LOG_DIRNAME, '', constants.ENV_LOG_DIRNAME, constants.DEFAULT_LOG_DIRNAME, (logDirname) => {
    if (typeof logDirname !== 'string') {
        return new Error(`Invalid log directory name ${logDirname}.`);
    }
    return (0, path_1.resolve)(logDirname);
});
const LogMaxFileSizeOption = new ConfigurableOption(constants.OPTIONS_MAX_LOG_FILE_SIZE, '', constants.ENV_MAX_LOG_FILE_SIZE, constants.DEFAULT_MAX_LOG_FILE_SIZE, (maxLogFileSize) => {
    if (typeof maxLogFileSize === 'string') {
        maxLogFileSize = parseInt(maxLogFileSize, 10);
    }
    if (isNaN(maxLogFileSize) || maxLogFileSize < 0) {
        return new Error(`Invalid max log file size ${maxLogFileSize}.`);
    }
    return maxLogFileSize;
});
const LogMaxFilesOption = new ConfigurableOption(constants.OPTIONS_MAX_LOG_FILES, '', constants.ENV_MAX_LOG_FILES, constants.DEFAULT_MAX_LOG_FILES, (maxLogFiles) => {
    if (typeof maxLogFiles === 'string') {
        maxLogFiles = parseInt(maxLogFiles, 10);
    }
    if (isNaN(maxLogFiles) || maxLogFiles < 0) {
        return new Error(`Invalid max log files ${maxLogFiles}.`);
    }
    return maxLogFiles;
});
const InterceptOutputOption = new ConfigurableOption(constants.OPTION_INTERCEPT_OUTPUT, '', constants.ENV_INTERCEPT_OUTPUT, constants.DEFAULT_INTERCEPT_OUTPUT, (interceptOutput) => {
    return typeof interceptOutput === 'string'
        && interceptOutput.toLowerCase() === 'true';
});
const ExtendedContextKeyOption = new ConfigurableOption(constants.OPTION_EXTENDED_CONTEXT_KEY, '', constants.ENV_EXTENDED_CONTEXT_KEY, constants.DEFAULT_EXTENDED_CONTEXT_KEY, (extendedContextKey) => {
    return typeof extendedContextKey === 'string' ? extendedContextKey : '';
});
const ExtendedContextOption = new ConfigurableOption(constants.OPTION_EXTENDED_CONTEXT, '', constants.ENV_EXTENDED_CONTEXT, constants.DEFAULT_EXTENDED_CONTEXT, (extendedContext) => {
    try {
        return typeof extendedContext === 'string' ? JSON.parse(extendedContext) : null;
    }
    catch (e) {
        logger_1.frameworkLogger.error(`Failed to parse extended context: '${extendedContext}', e: ${e}`, extendedContext, e);
        throw e;
    }
});
const EnableCorsOption = new ConfigurableOption(constants.OPTION_ENABLE_CORS, '', constants.ENV_ENABLE_CORS, constants.DEFAULT_ENABLE_CORS, (enableCors) => {
    if (typeof enableCors === 'string') {
        enableCors = enableCors.toLowerCase() === 'true';
    }
    return enableCors;
});
const AllowedOriginsOption = new ConfigurableOption(constants.OPTION_ALLOWED_ORIGINS, '', constants.ENV_ALLOWED_ORIGINS, [], (allowedOrigins) => {
    return typeof allowedOrigins === 'string' ? allowedOrigins.split(',').map(origin => origin.trim()) : [];
});
const AllowedAccounts = new ConfigurableOption(constants.OPTION_ALLOWED_ACCOUNTS, '', constants.ENV_ALLOWED_ACCOUNTS, [], (allowedAccounts) => {
    return typeof allowedAccounts === 'string' ? allowedAccounts.split(',').map(item => {
        const [username, password] = item.split(':');
        return { username, password };
    }) : [];
});
const StaticServeRoot = new ConfigurableOption(constants.OPTION_STATIC_SERVE_ROOT, '', constants.ENV_STATIC_SERVE_ROOT, constants.DEFAULT_STATIC_SERVE_ROOT, (staticServeRoot) => staticServeRoot);
const StaticServeIndex = new ConfigurableOption(constants.OPTION_STATIC_SERVE_INDEX, '', constants.ENV_STATIC_SERVE_INDEX, constants.DEFAULT_STATIC_SERVE_INDEX, (staticServeIndex) => staticServeIndex);
const StaticServeMaxAge = new ConfigurableOption(constants.OPTION_STATIC_SERVE_MAX_AGE, '', constants.ENV_STATIC_SERVE_MAX_AGE, constants.DEFAULT_STATIC_SERVE_MAX_AGE, (staticServeMaxAge) => {
    if (typeof staticServeMaxAge === 'string') {
        staticServeMaxAge = parseInt(staticServeMaxAge, 10);
    }
    if (isNaN(staticServeMaxAge)) {
        return new Error(`Invalid staticServeMaxAge:${staticServeMaxAge}.`);
    }
    return staticServeMaxAge;
});
const DotEnvFilePathOption = new ConfigurableOption(constants.DOTENV_FILE_PATH, '', constants.ENV_DOTENV_FILE_PATH, constants.DEFAULT_DOTENV_FILE_PATH, (dotEnvFilePath) => dotEnvFilePath);
const FunctionsConfigOption = new ConfigurableOption(constants.OPTIONS_FUNCTIONS_CONFIG, '', constants.ENV_FUNCTIONS_CONFIG, constants.DEFAULT_FUNCTIONS_CONFIG, (multiFunctionsConfig) => {
    if (typeof multiFunctionsConfig !== 'string') {
        return new Error(`Invalid multi-functions config ${multiFunctionsConfig}.`);
    }
    return multiFunctionsConfig;
});
exports.helpText = `
Usage: tcb-ff [options]

Options:
  -w, --watch               Watch the source file for changes and restart the service. Default is false.
                            Example: -w

  --port, -p                Set the port to listen on. Default is ${constants.DEFAULT_PORT}.
                            Example: --port=3000
                            Environment Variable: PORT

  --target                  Set the target function name. Default is '${constants.DEFAULT_TARGET}'.
                            Example: --target=hello
                            Environment Variable: FUNCTION_TARGET

  --source                  Set the target function filename. Default is 'index.js' (extension differs based on module type).
                            Example: --source=main.mjs
                            Environment Variable: FUNCTION_SOURCE

   --loadAllFunctions       Whether to load all functions in the 'functionsRoot' directory. Default: false.
                            Example: --loadAllFunctions=true
                            Environment Variable: LOAD_ALL_FUNCTIONS

   --functionsRoot          The root directory of the functions. only effective when 'loadAllFunctions' is true.
                            Example: --functionsRoot=path/to/functionsRoot
                            Environment Variable: FUNCTIONS_ROOT

  --signatureType           Set the function format. Currently only 'event' is supported.
                            Example: --signatureType=event
                            Environment Variable: SIGNATURE_TYPE

  --gracefulShutdown        Enable graceful shutdown of the service. Default true when 'NODE_ENV=production' otherwise false.
                            Example: --gracefulShutdown=false
                            Environment Variable: GRACEFUL_SHUTDOWN

  --timeout                 Set the Node.js Server timeout. Default is ${constants.DEFAULT_TIMEOUT}.
                            Example: --timeout=5000
                            Environment Variable: TIMEOUT

  --keepAliveTimeout        Set the Node.js Server keep-alive timeout. Default is ${constants.DEFAULT_KEEP_ALIVE_TIMEOUT}.
                            Example: --keepAliveTimeout=10000
                            Environment Variable: KEEP_ALIVE_TIMEOUT

  --captureStack            Return error stack when a request fails. Enabled when 'NODE_ENV=production'.
                            Example: --captureStack=true
                            Environment Variable: CAPTURE_STACK

  --concurrencyLimit        Limit the number of concurrent requests. 0 means no limit. Default is ${constants.DEFAULT_CONCURRENCY_LIMIT}.
                            Example: --concurrencyLimit=100
                            Environment Variable: CONCURRENCY_LIMIT

  --noExit                  Disable 'process.exit()' to prevent accidental process exit. Default is ${constants.DEFAULT_NO_EXIT}.
                            Example: --noExit=true
                            Environment Variable: NO_EXIT

  --interceptOutput         If true, intercept console、process.stdout、process.stderr output. Default: false.
                            Example: --interceptOutput=true
                            Environment Variable: INTERCEPT_OUTPUT

  --logHeaderBody           Log request and response headers and bodies in the request log. 
                            Default is ${constants.DEFAULT_LOG_HEADER_BODY}.
                            Example: --logHeaderBody=false
                            Environment Variable: LOG_HEADER_BODY

  --logEventContext         Log original event and context in the access log. Default is ${constants.DEFAULT_LOG_EVENT_CONTEXT}.
                            Example: --logEventContext=false
                            Environment Variable: LOG_EVENT_CONTEXT

  --logContentEnabled       Include simplified __CONTENT__ field in logs. Default is ${constants.DEFAULT_LOG_CONTENT_ENABLED}.
                            Example: --logContentEnabled=false
                            Environment Variable: LOG_CONTENT_ENABLED

  --extendedContextKey      The key of headers to parse \`context.extendedContext\`.
                              \`''\` means that the feature is turned off. 
                            Default is ${constants.DEFAULT_EXTENDED_CONTEXT_KEY}.
                            Example: --extendedContextKey=X-Functions-Extended-Context
                            Environment Variable: EXTENDED_CONTEXT_KEY

  --extendedContext         The value to parse \`context.extendedContext\`.
                            \`''\` means that the feature is turned off.
                            This parameter will be ignore when 'extendedContextKey' is specified.
                            Default is ${constants.DEFAULT_EXTENDED_CONTEXT}.
                            Example: --extendedContext='{"a":1,"b":2}'
                            Environment Variable: EXTENDED_CONTEXT

  --functionsConfigFile     Set the multi-functions definition config file. Default is '${constants.DEFAULT_FUNCTIONS_CONFIG}'.
                            Example: --functionsConfigFile=functions.json
                            Environment Variable: FUNCTIONS_CONFIG_FILE

  --enableCors              If true, enable CORS for configured origins. Default is ${constants.DEFAULT_ENABLE_CORS}.
                            Example: --enableCors=true
                            Environment Variable: ENABLE_CORS

  --allowedOrigins          Set the allowed origins for CORS. Default allows ${constants.DEFAULT_ALLOWED_ORIGINS}.
                            Origins should be domain names without protocol or port.
                            Wildcard origins are supported. e.g. ['*.example.com'].
                            Multiple origins should be separated by comma.
                            Example: --allowedOrigins=*.example.com,www.another.com
                            Environment Variable: ALLOWED_ORIGINS

  --allowedAccounts         Set the accounts allowed to access.
                            If specified, The 'username:password' of request will be verified.
                            Parameter Format: username1:password1,username1:password1,....
                            Example: --allowedAccounts=abc:123,efg:456
                            Access Example: 
                              curl http://username:password@127.0.0.1:3000/
                              curl -H "Authorization: Basic base64encode(username:password)" http://127.0.0.1:3000/
                            Environment Variable: ALLOWED_USERS

  --staticServeRoot         Set root directory of static file.
                            Example: --staticServeRoot=public
                            Environment Variable: STATIC_SERVE_ROOT

  --staticServeIndex        Set index of static file.
                            Example: --staticServeRoot=index.html
                            Environment Variable: STATIC_SERVE_INDEX

  --staticServeMaxAge       Set MaxAge of static file in milliseconds. Default 0.
                            Example: --staticServeMaxAge=3600
                            Environment Variable: STATIC_SERVE_MAX_AGE

  --dotEnvFilePath        The path to the .env file. Unspecified means off. Support multiple paths separated by ','.
                            Example: --dotEnvFilePath=path/to/.env,path/to/.env.local

  --dry-run                 Do not start the service, only validate that the code can be loaded. Default is false.
                            Example: --dry-run
`;
const parseOptions = (cliArgs = process.argv, envVars = process.env) => {
    const argv = (0, minimist_1.default)(cliArgs, {
        string: [
            PortOption.inputOption,
            FunctionTargetOption.inputOption,
            FunctionSourceOption.inputOption,
            FunctionsRootOption.inputOption,
            enableLoadAllFunctionsOption.inputOption,
            SignatureOption.inputOption,
            TimeoutOption.inputOption,
            GracefulShutDownOption.inputOption,
            KeepAliveTimeoutOption.inputOption,
            ConcurrencyLimitOption.inputOption,
            NoExitOption.inputOption,
            InterceptOutputOption.inputOption,
            CaptureStackTrace.inputOption,
            LogHeaderBodyOption.inputOption,
            LogDirnameOption.inputOption,
            LogMaxFileSizeOption.inputOption,
            LogMaxFilesOption.inputOption,
            LogEventContextOption.inputOption,
            LogContentEnabledOption.inputOption,
            ExtendedContextKeyOption.inputOption,
            FunctionsConfigOption.inputOption,
            EnableCorsOption.inputOption,
            AllowedOriginsOption.inputOption,
            DotEnvFilePathOption.inputOption,
            AllowedAccounts.inputOption
        ],
        alias: {
            [PortOption.inputOption]: PortOption.alias,
            [SignatureOption.inputOption]: SignatureOption.alias
        }
    });
    return {
        dotEnvFilePath: DotEnvFilePathOption.parse(argv, envVars),
        port: PortOption.parse(argv, envVars),
        functionsRoot: FunctionsRootOption.parse(argv, envVars),
        loadAllFunctions: enableLoadAllFunctionsOption.parse(argv, envVars),
        target: FunctionTargetOption.parse(argv, envVars),
        sourceLocation: FunctionSourceOption.parse(argv, envVars),
        signatureType: SignatureOption.parse(argv, envVars),
        timeoutMS: TimeoutOption.parse(argv, envVars),
        gracefulShutdown: GracefulShutDownOption.parse(argv, envVars),
        keepAliveTimeoutMS: KeepAliveTimeoutOption.parse(argv, envVars),
        concurrencyLimit: ConcurrencyLimitOption.parse(argv, envVars),
        noExit: NoExitOption.parse(argv, envVars),
        interceptOutput: InterceptOutputOption.parse(argv, envVars),
        captureStack: CaptureStackTrace.parse(argv, envVars),
        logHeaderBody: LogHeaderBodyOption.parse(argv, envVars),
        logEventContext: LogEventContextOption.parse(argv, envVars),
        logContentEnabled: LogContentEnabledOption.parse(argv, envVars),
        dryRun: cliArgs.includes(constants.OPTION_DRY_RUN),
        logDirname: LogDirnameOption.parse(argv, envVars),
        maxLogFileSize: LogMaxFileSizeOption.parse(argv, envVars),
        maxLogFiles: LogMaxFilesOption.parse(argv, envVars),
        extendedContextKey: ExtendedContextKeyOption.parse(argv, envVars),
        extendedContext: ExtendedContextOption.parse(argv, envVars),
        functionsConfigFile: FunctionsConfigOption.parse(argv, envVars),
        enableCors: EnableCorsOption.parse(argv, envVars),
        allowedOrigins: AllowedOriginsOption.parse(argv, envVars),
        allowedAccounts: AllowedAccounts.parse(argv, envVars),
        staticServeRoot: StaticServeRoot.parse(argv, envVars),
        staticServeIndex: StaticServeIndex.parse(argv, envVars),
        staticServeMaxAge: StaticServeMaxAge.parse(argv, envVars),
        printHelp: cliArgs[2] === '--help' || cliArgs[2] === '-h'
    };
};
exports.parseOptions = parseOptions;
