// @cloudbase/functions-framework/build/router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Router = void 0;
exports.routeFunction = routeFunction;
const radix3_1 = require("radix3");
const function_loader_1 = require("./function-loader");
const function_registry_1 = require("./function-registry");
class Router {
    routes = new Map();
    router;
    constructor(options) {
        this.router = (0, radix3_1.createRouter)({
            // 不严格检查路径末尾的斜杠，例如 /foo 和 /foo/ 会被认为是同一个路由
            strictTrailingSlash: false
        });
        if (!(options && options.routes)) {
            return;
        }
        // 将 routes 中的 path 格式化，认为 /aa 和 /aa/ 相同
        for (const route of options.routes) {
            if (route.path) {
                const p = route.path.endsWith('/') ? route.path : route.path + '/';
                if (this.routes.has(p)) {
                    continue;
                }
                this.routes.set(p, route);
                this.routes.set(p + '**', route); // 添加前缀匹配路由
            }
        }
        for (const [path, route] of this.routes.entries()) {
            this.addRoute(path, route);
        }
    }
    addRoute(path, route) {
        this.router.insert(path, route);
    }
    /**
     * 根据 path 查找路由信息，返回匹配到的原始路由定义
     * @param path 路径
     * @returns 路由信息
     */
    findRoute(path) {
        const found = this.router.lookup(path);
        if (!found) {
            return null;
        }
        if (found.params) {
            delete found.params;
        }
        return found;
    }
}
exports.Router = Router;
/**
 * 根据 path 查找匹配的路由。
 *
 * 路由规则：
 * 1. 每一条路由的 path 匹配均为前缀匹配，即 /aa 可以匹配 /aa/bb ，/aa/bb/cc等等
 * 2. 如有多条相同 path 规则，先声明的优先匹配
 * 3. path 规则中带 / 的和不带 / 的路由视为同一条路由，如 /aa 和 /aa/ 认为是相同的，前面的规则会覆盖后面的，即以第一条声明的为准
 *     a. 如果请求时 path 为空，将会被视作 / ，会命中配置 / 的规则（如果未配置 / ，则 404）
 * 4. 如有多条规则匹配，将会命中最长匹配的规则，如 /a -> a , /a/b -> b ，path /a/b/c将会匹配至 b
 * 5. 匹配以 / 分隔为最小的一段，单个段内不会模糊匹配，如/aaa 不会匹配 /a -> a 规则
 *
 * @param router IRouter 实例
 * @param path 路径，不带有 query 和 hash
 * @param option 配置项
 * @returns
 */
function routeFunction(router, fnName, path) {
    if (router === undefined) {
        const fnId = (0, function_registry_1.buildFnId)(fnName, function_loader_1.DEFAULT_FN_NAME);
        return {
            handleFunction: (0, function_registry_1.getRegisteredFunction)(fnId)
        };
    }
    const matchedRoute = router.findRoute(path);
    if (matchedRoute) {
        return {
            routeDefinition: matchedRoute,
            handleFunction: (0, function_registry_1.getRegisteredFunction)((0, function_registry_1.buildFnId)(fnName, matchedRoute.functionName))
        };
    }
    return {};
}
