// @cloudbase/functions-framework/build/server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createServer = createServer;
const http_1 = __importDefault(require("http"));
const koa_1 = __importDefault(require("koa"));
const koa_body_1 = require("koa-body");
const koa_static_1 = __importDefault(require("koa-static"));
const raw_body_1 = __importDefault(require("raw-body"));
const error_1 = require("./error");
const accesslog_1 = require("./middlewares/accesslog");
const async_context_1 = require("./middlewares/async-context");
const basic_auth_1 = require("./middlewares/basic-auth");
const catch_error_1 = require("./middlewares/catch-error");
const concurrency_limit_1 = require("./middlewares/concurrency-limit");
const context_1 = require("./middlewares/context");
const cors_1 = require("./middlewares/cors");
const eventid_1 = require("./middlewares/eventid");
const function_route_1 = require("./middlewares/function-route");
const healthz_1 = require("./middlewares/healthz");
const timeout_1 = require("./middlewares/timeout");
const unified_responder_1 = require("./unified-responder");
const wss_1 = require("./wss");
function createServer(options, routes) {
    const app = new koa_1.default();
    app.routers = routes;
    app.use((0, healthz_1.healthzMiddleware)());
    app.use((0, basic_auth_1.basicAuthMiddleware)(options.allowedAccounts));
    app.use((0, eventid_1.eventidMiddleware)());
    app.use((0, async_context_1.asyncContextMiddleware)());
    app.use((0, accesslog_1.accessLogMiddleware)({
        logHeaderBody: options.logHeaderBody,
        logEventContext: options.logEventContext,
        logContentEnabled: options.logContentEnabled
    }));
    app.use((0, catch_error_1.catchErrorMiddleware)());
    app.use((0, timeout_1.timeoutMiddleware)(options.timeoutMS));
    app.use((0, concurrency_limit_1.concurrencyLimitMiddleware)(options.concurrencyLimit));
    app.use((0, cors_1.corsMiddleware)(options.enableCors, options.allowedOrigins));
    // 静态文件服务，注意：先尝试进行静态文件匹配，如无匹配再向后进行匹配
    if (options.staticServeRoot !== undefined) {
        app.use((0, koa_static_1.default)(options.staticServeRoot, {
            index: options.staticServeIndex || false,
            maxAge: options.staticServeMaxAge || 0,
            hidden: false,
            defer: false
        }));
    }
    // attach body parser middleware
    app.use((0, koa_body_1.koaBody)({
        multipart: true,
        patchKoa: true,
        jsonLimit: '1024mb',
        formLimit: '1024mb',
        textLimit: '1024mb',
        formidable: {
            maxFileSize: 1024 * 1024 * 1024, // 1GB
            maxTotalFileSize: 1024 * 1024 * 1024 // 1GB
        },
        onError: (err, ctx) => {
            return (0, unified_responder_1.sendResponse)(ctx, undefined, (0, error_1.newBadRequestErr)(err), 400);
        }
    }));
    app.use(async (ctx, next) => {
        // if the req body has been parsed, avoid reading it again.
        // This is closely related to the behavior of koa-body
        if (!(ctx.request.files || ctx.request.body || ctx.state.isTimeout)) {
            try {
                ctx.rawBody = await (0, raw_body_1.default)(ctx.req);
            }
            catch (err) {
                return (0, unified_responder_1.sendResponse)(ctx, undefined, (0, error_1.newBadRequestErr)(err), 400);
            }
        }
        await next();
    });
    app.use((0, context_1.contextInjectionMiddleware)(options.extendedContextKey, options.extendedContext));
    app.use((0, function_route_1.functionRouteMiddleware)(app.routers, options));
    const httpServer = http_1.default.createServer(app.callback());
    (0, wss_1.createWebSocketServer)(httpServer, app, options);
    return httpServer;
}
