// @cloudbase/functions-framework/build/tools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerSignalHandler = exports.killInstance = void 0;
exports.interceptProcessExit = interceptProcessExit;
exports.isProductionEnv = isProductionEnv;
exports.getCaller = getCaller;
exports.getFormatedStackTraces = getFormatedStackTraces;
exports.formatCallSite = formatCallSite;
exports.getStackTraces = getStackTraces;
const constants_1 = require("./constants");
const error_1 = require("./error");
const logger_1 = require("./logger");
const options_1 = require("./options");
const user_code_log_1 = require("./user-code-log");
// The two bind will happen before intercept process.exit so they
// will persist the correct context.
exports.killInstance = process.exit.bind(process, 16); // special code for framework
const stopInstance = (err) => process.exit.bind(process, err ? 1 : 0);
/**
 * ServerSignalHandler 处理进程相关信号
 */
class ServerSignalHandler {
    server;
    killTimeout;
    gracefulShutdown;
    constructor(server, gracefulShutdown = true) {
        this.server = server;
        this.killTimeout = new options_1.ConfigurableOption('', '', constants_1.ENV_KILL_TIMEOUT, constants_1.DEFAULT_KILL_TIMEOUT)
            .parse([], process.env);
        this.gracefulShutdown = gracefulShutdown;
    }
    setup() {
        // set up process signal handlers
        process.on('uncaughtException', (err, origin) => {
            logger_1.frameworkLogger.error(`${origin}: ${(0, error_1.decorateErrorStack)(err)}`);
        });
        process.on('unhandledRejection', (reason) => {
            logger_1.frameworkLogger.error(`Unhandled Rejection for: ${(0, error_1.decorateErrorStack)(reason)}`);
        });
        process.on('exit', (code) => {
            logger_1.frameworkLogger.info(`Process exited with code: ${code}`);
        });
        ['SIGINT', 'SIGTERM'].forEach((signal) => {
            process.on(signal, () => {
                if (!this.gracefulShutdown) {
                    logger_1.frameworkLogger.info(`Server received ${signal}, closing now.`);
                    setImmediate(stopInstance(null));
                    return;
                }
                logger_1.frameworkLogger.info(`Server received ${signal}, will close after connections are drained.`);
                this.server.close((err) => {
                    setImmediate(() => stopInstance(err));
                });
                if (process.env.NODE_ENV === 'production') {
                    setTimeout(() => {
                        this.server.getConnections((err) => {
                            if (err) {
                                logger_1.frameworkLogger.error(`Error waiting for connections to close: ${err}`);
                                stopInstance(err);
                            }
                        });
                        setImmediate(exports.killInstance);
                    }, this.killTimeout);
                }
            });
        });
    }
}
exports.ServerSignalHandler = ServerSignalHandler;
process.forceExit = process.exit.bind(process);
/**
 * Intercept process.exit calls to avoid unexpected exits
 * @returns
 */
function interceptProcessExit() {
    const exitHandler = {
        apply: () => {
            const tips = '[WARN]Operation aborted: `process.exit` calls have been disabled in this context. No action will be taken.';
            // 注意：重构代码注意 caller 变化
            (0, user_code_log_1.addUserCodeLog)({ level: 'warn', content: tips, caller: getCaller(null, 2) });
            logger_1.frameworkLogger.error(tips);
        }
    };
    process.exit = new Proxy(process.exit, exitHandler);
}
/**
 * 是否是生产环境
 * @returns
 */
function isProductionEnv() { return process.env.NODE_ENV === 'production'; }
function getCaller(belowFn = null, index = 1) {
    const stackTraces = getStackTraces(belowFn);
    return formatCallSite(stackTraces[index]);
}
function getFormatedStackTraces(belowFn = null) {
    const stackTraces = getStackTraces(belowFn);
    return stackTraces.map(formatCallSite).join('\n');
}
function formatCallSite(stackTrace) {
    return `at ${stackTrace.getFunctionName() || '<anonymous>'} (${stackTrace.getFileName()}:${stackTrace.getLineNumber()}:${stackTrace.getColumnNumber()})`;
}
function getStackTraces(belowFn = null) {
    const originalPrepareStackTrace = Error.prepareStackTrace;
    Error.prepareStackTrace = function (dummyObject, stackTraces) { return stackTraces; };
    const dummyObject = { stack: [] };
    Error.captureStackTrace(dummyObject, belowFn || getStackTraces);
    const stackTraces = dummyObject.stack;
    Error.prepareStackTrace = originalPrepareStackTrace;
    return stackTraces;
}
