// @cloudbase/functions-framework/build/unified-responder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendResponse = sendResponse;
const error_1 = require("./error");
function sendResponse(ctx, result, err, errStatusCode = 500, withStack = false) {
    if (err) {
        if (!ctx.response) {
            return;
        }
        ctx.status = errStatusCode;
        if (withStack) {
            result = {
                code: err.code,
                message: err.message,
                stack: (0, error_1.decorateErrorStack)(err.cause),
                requestId: ctx.state.eventID
            };
        }
        else {
            result = {
                code: err.code,
                message: err.message,
                requestId: ctx.state.eventID
            };
        }
    }
    if (isIntegrationResponse(result)) {
        return sendIntegrationResponse(ctx, result);
    }
    if (typeof result === 'undefined' || result === null) {
        ctx.status = 204;
    }
    else {
        ctx.body = result;
    }
    const eResult = result;
    if (eResult && eResult.code) {
        ctx.state.errorResult = {
            code: eResult.code,
            message: eResult.message
        };
    }
    return;
}
const isIntegrationResponse = (result) => {
    if (typeof result !== 'object' || !result) {
        return false;
    }
    const { statusCode, headers } = result;
    return typeof statusCode === 'number' && typeof headers === 'object';
};
function sendIntegrationResponse(ctx, result) {
    ctx.set(result.headers);
    ctx.body = result.body;
    // 设置 status 应放在最后，否则可能被 Koa 默认的 204 等覆盖
    ctx.status = result.statusCode;
}
