// @cloudbase/functions-framework/build/user-code-log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.takeUserCodeLogs = takeUserCodeLogs;
exports.pushUserCodeLogs = pushUserCodeLogs;
exports.addUserCodeLog = addUserCodeLog;
const async_context_1 = require("./async-context");
const constants_1 = require("./constants");
const logger_1 = require("./logger");
function lazyInitUserCodeLogs() {
    const currentContext = async_context_1.asyncContext.getStore();
    if (currentContext && currentContext.userCodeLogs === undefined) {
        currentContext.userCodeLogs = [];
    }
}
/**
 * 取走当前异步上下文关联的日志
 * @returns
 */
function takeUserCodeLogs() {
    lazyInitUserCodeLogs();
    const context = (0, async_context_1.getCurrentAsyncContext)();
    if (context !== undefined) {
        const logs = context.userCodeLogs || [];
        context.userCodeLogs = [];
        return logs;
    }
    return [];
}
/**
 * 添加上下文关联的日志
 * @returns
 */
function pushUserCodeLogs(log) {
    lazyInitUserCodeLogs();
    const context = (0, async_context_1.getCurrentAsyncContext)();
    if (context?.userCodeLogs) {
        context.userCodeLogs.push(log);
    }
}
/**
 * 添加一条日志到当前异步上下文
 * @param content 日志内容
 * @param caller 调用处
 */
function addUserCodeLog(log) {
    lazyInitUserCodeLogs();
    if (!log['@timestamp']) {
        log['@timestamp'] = new Date().toISOString();
    }
    if (!log['level']) {
        log['level'] = 'log';
    }
    if (!log['caller']) {
        // log['caller'] = getCaller(addUserCodeLog)
        // log['stacktraces'] = getFormatedStackTraces()
    }
    // 如果函数已经返回，日志两种刷新方式：
    // 1. 日志条数达到限制值触发刷新
    // 2. 定时器触发刷新
    const userCodeLogs = (0, async_context_1.getCurrentAsyncContext)()?.userCodeLogs || [];
    if (userCodeLogs.length >= constants_1.USER_CODE_LOGS_MAX_COUNT) {
        logger_1.functionsLogger.logUserCodelog();
    }
    {
        const currentContext = (0, async_context_1.getCurrentAsyncContext)();
        if (currentContext && currentContext.done && !currentContext.userCodeLogsFlushTimer) {
            currentContext.userCodeLogsFlushTimer = setTimeout(() => {
                logger_1.functionsLogger.logUserCodelog();
                currentContext.userCodeLogsFlushTimer = undefined;
            }, constants_1.USER_CODE_LOGS_FLUSH_TIMER_TIMEOUT);
        }
    }
    pushUserCodeLogs(log);
}
