// @cloudbase/functions-framework/build/wss.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWebSocketServer = createWebSocketServer;
const events = __importStar(require("events"));
const http = __importStar(require("http"));
const koa_compose_1 = __importDefault(require("koa-compose"));
const ws_1 = require("ws");
const error_1 = require("./error");
const function_registry_1 = require("./function-registry");
const logger_1 = require("./logger");
const accesslog_1 = require("./middlewares/accesslog");
const async_context_1 = require("./middlewares/async-context");
const context_1 = require("./middlewares/context");
const eventid_1 = require("./middlewares/eventid");
const router_1 = require("./router");
const utils_lang_1 = require("./utils.lang");
const WS_ERROR_CODE = {
    SERVICE_RESTART: 1012
};
/**
 * respond error to http request
 * @param socket
 * @param e
 */
function respondError(socket, e) {
    respond(socket, 500, `Internal Server Error: ${e && e.message ? e.message : e}`);
}
/**
 * respond to http request
 *
 * 在处理过程中，可能出现某些状况，这里通过 HTTP 的方式回复客户端，方便从客户端侧排查问题，部分客户端工具（如curl）可以看到服务端响应内容
 *
 * 一旦 http 协议进入 websocket 握手阶段，就不再是 http 协议，也就没有 response 对象，所以这里需要手动构造 http 响应
 *
 * @param socket
 * @param statusCode
 * @param body
 */
function respond(socket, statusCode, body, type = 'text/plain') {
    // 不允许升级给出 http 响应，这里默认给 404，可由业务自定义
    const statusMessage = http.STATUS_CODES[statusCode];
    const head = [
        `HTTP/1.1 ${statusCode} ${statusMessage}`,
        `Content-Type: ${type}`,
        `Content-Length: ${Buffer.byteLength(Buffer.from(body))}`,
        'Content-Encoding: utf-8',
        'Connection: close'
    ].join('\r\n');
    socket.write(head);
    socket.write('\r\n\r\n');
    socket.end(body);
}
/**
 * setupWebSocketsHeartbeat
 * @param wss
 * @param interval
 */
function setupWebSocketsHeartbeat(wss, interval = 10000) {
    const SYM_WS_IS_ALIVE = Symbol('WS_IS_ALIVE');
    let checking = false;
    let wsNotAliveCount = 0;
    function doCheck() {
        if (checking) {
            return;
        }
        checking = true;
        wsNotAliveCount = 0;
        // 通过 ping-pong 机制检查存活
        wss.clients.forEach((ws) => {
            if (Reflect.has(ws, SYM_WS_IS_ALIVE) === false) {
                // 仅首次探测执行一次
                ws[SYM_WS_IS_ALIVE] = true;
                ws.on('pong', function onPong() {
                    this[SYM_WS_IS_ALIVE] = true;
                });
            }
            if (ws.readyState !== ws.OPEN) {
                return;
            }
            if (ws[SYM_WS_IS_ALIVE] === false) {
                wsNotAliveCount += 1;
                ws.close(1001, 'websocket client not alive');
                return ws.terminate();
            }
            ws[SYM_WS_IS_ALIVE] = false;
            ws.ping();
        });
        logger_1.frameworkLogger.info(`wss heartbeat finish, ws not alive count: ${wsNotAliveCount}, total: ${wss.clients.size}`);
        checking = false;
    }
    const aliveCheckTimer = setInterval(doCheck, interval);
    wss.once('close', () => {
        clearInterval(aliveCheckTimer);
    });
}
/**
 * setupGracefulShutdown
 * @param wss
 */
function setupGracefulShutdown(wss) {
    ['SIGINT', 'SIGTERM'].forEach((signal) => {
        process.on(signal, async () => {
            logger_1.frameworkLogger.info(`received ${signal} signal, shutting down websocket server, ${wss.clients.size} ws clients need to be closed`);
            wss.close();
            wss.clients.forEach((ws) => {
                ws.close(WS_ERROR_CODE.SERVICE_RESTART, 'websocket server is shutting down, please try to reconnect');
            });
        });
    });
}
/**
 * createWebSocketServer 创建 WebSocket 服务，完成 websocket 握手后，调用用户函数
 * @param httpServer
 * @param koaApp
 * @param userFunction
 * @returns
 */
function createWebSocketServer(httpServer, koaApp, options) {
    const hasExportHandleUpgradeFunctions = [];
    (0, function_registry_1.getAllRegisteredFunctions)().forEach(fn => {
        if (typeof fn.userFunction.handleUpgrade === 'function') {
            hasExportHandleUpgradeFunctions.push(fn.name);
        }
    });
    // 关于 userFunction.handleUpgrade 函数：
    // 该函数用于判断是否能够升级协议，同时也作为是否开启 WebSocket 的标志
    // 这里设计为在 `入口函数` 上挂一个 `handleUpgrade` 方法，用于处理 WebSocket 的握手，例如认证鉴权等
    // 为什么在 `入口函数` 上挂一个 函数？简化函数加载的过程，不需要额外的函数加载
    // 为什么不直接在 `入口函数` 中处理 WebSocket 握手？明确区分 `协议升级` 和 `消息收发` 过程，简化 `入口函数` 的逻辑
    if (hasExportHandleUpgradeFunctions.length === 0) {
        return;
    }
    logger_1.frameworkLogger.info('WebSocket feature is enabled. Enabled functions: ', hasExportHandleUpgradeFunctions.join(', '));
    const wss = new ws_1.WebSocketServer({
        noServer: true,
        clientTracking: true,
        perMessageDeflate: {
            zlibDeflateOptions: {
                // See zlib defaults.
                chunkSize: 1024,
                memLevel: 7,
                level: 3
            },
            zlibInflateOptions: {
                chunkSize: 10 * 1024
            },
            // Other options settable:
            clientNoContextTakeover: true, // Defaults to negotiated value.
            serverNoContextTakeover: true, // Defaults to negotiated value.
            serverMaxWindowBits: 10, // Defaults to negotiated value.
            // Below options specified as default values.
            concurrencyLimit: 10, // Limits zlib concurrency for perf.
            threshold: 1024 // Size (in bytes) below which messages
            // should not be compressed if context takeover is disabled.
        }
    });
    setupGracefulShutdown(wss);
    setupWebSocketsHeartbeat(wss);
    wss.on('error', err => {
        logger_1.frameworkLogger.error('ws server error: ', err);
    });
    wss.on('close', () => {
        logger_1.frameworkLogger.error('ws server has been closed');
    });
    // 握手过程中，将要发送的 headers，这里可以修改 headers，如添加 RequestId，这里暂不处理
    // wss.on('headers', (headers: string[], req: http.IncomingMessage) => {})
    // 握手完成时触发，noServer:false 握手完成会触发
    // wss.on('connection', () => {})
    // HttpServer 监听端口时触发，noServer:false 时有效
    // wss.on('listening', () => {})
    // 这里复用 koa 中间件机制，以便复用中间件的功能
    const composedMiddleware = (0, koa_compose_1.default)([
        (0, eventid_1.eventidMiddleware)(),
        (0, async_context_1.asyncContextMiddleware)(),
        (0, accesslog_1.accessLogMiddleware)({
            logHeaderBody: options.logHeaderBody,
            logEventContext: options.logEventContext,
            logContentEnabled: options.logContentEnabled
        }),
        async (ctx, next) => {
            try {
                await next();
            }
            catch (err) {
                logger_1.frameworkLogger.error('catch error: ', err);
            }
        },
        (0, context_1.contextInjectionMiddleware)(options.extendedContextKey, options.extendedContext)
    ]);
    // 这里在未开启 WebSocket 的情况下，依然监听 upgrade 事件
    // 防止 websocket upgrade 请求被当做普通 http 请求处理
    httpServer.on('upgrade', async (request, socket, head) => {
        let fnName = '';
        if (options.loadAllFunctions) {
            const url = new URL(`http://127.0.0.1${request.url || ''}`);
            fnName = url.searchParams.get('@fn') || '';
        }
        const router = koaApp.routers.get(fnName);
        const { routeDefinition, handleFunction } = (0, router_1.routeFunction)(router, fnName, request.url || '/');
        const userFunction = handleFunction?.userFunction;
        if (typeof userFunction === 'undefined' || typeof userFunction.handleUpgrade !== 'function') {
            socket.destroy();
            return;
        }
        request.socket.setKeepAlive(true);
        request.socket.setNoDelay(true);
        request.socket.setTimeout(0);
        try {
            // 注意：这里手动创建一个 ctx，和 koa 内部创建的 ctx 主要是 res 参数不同，中间件实现需要兼容可能存在的差异
            /* eslint-disable-next-line */
            const ctx = koaApp.createContext(request, {});
            if (routeDefinition !== null) {
                ctx.state.matchedRoute = routeDefinition;
            }
            await composedMiddleware(ctx, () => {
                return new Promise((resolve) => {
                    // 这里没啥用，主要解决 ts 类型报错
                    if (typeof userFunction.handleUpgrade !== 'function') {
                        respond(socket, 404, 'Not Found');
                        return;
                    }
                    // Upgrade 过程的上下文
                    const upgradeContext = {
                        ctxId: ctx.state.eventID,
                        eventID: ctx.state.eventID,
                        eventType: 'websocket.upgrade',
                        timestamp: new Date().toISOString(),
                        httpContext: {
                            url: request.url,
                            httpMethod: request.method,
                            headers: request.headers
                        }
                    };
                    if (Object.keys(ctx.extendedContext || {}).length > 0) {
                        upgradeContext.extendedContext = ctx.extendedContext;
                    }
                    ctx.state.contextInjected = upgradeContext;
                    // ctx.state.event = {}
                    try {
                        Promise.resolve(userFunction.handleUpgrade(upgradeContext)).then((result) => {
                            if (!result.allowWebSocket) {
                                respond(socket, result.statusCode || 404, result.body || 'Not Found', result.contentType);
                                return;
                            }
                            wss.handleUpgrade(request, socket, head, async (ws) => {
                                const context = {
                                    ctxId: ctx.state.eventID,
                                    eventID: ctx.state.eventID,
                                    eventType: 'websocket.connection',
                                    timestamp: new Date().toISOString(),
                                    extendedContext: ctx.extendedContext,
                                    httpContext: {
                                        url: request.url,
                                        httpMethod: request.method,
                                        headers: request.headers
                                    }
                                };
                                context.ws = new WebSocket(ws, { ctxId: context.ctxId });
                                (0, utils_lang_1.deepFreeze)(context);
                                process.nextTick(() => {
                                    try {
                                        userFunction(null, { ...context });
                                    }
                                    catch (e) {
                                        // 忽略错误，如果需要打日志，用户自己 try-catch
                                    }
                                });
                                resolve(undefined);
                            });
                        }).catch((e) => {
                            logger_1.frameworkLogger.error('handleUpgrade promise-catch error: ', (0, error_1.decorateErrorStack)(e));
                            respondError(socket, e);
                        });
                    }
                    catch (e) {
                        logger_1.frameworkLogger.error('handleUpgrade try-catch error: ', (0, error_1.decorateErrorStack)(e));
                        respondError(socket, e);
                    }
                });
            });
        }
        catch (e) {
            logger_1.frameworkLogger.error('handleUpgrade error: ', e);
            respondError(socket, e);
        }
    });
}
/**
 * WebSocket 主要是对 ws.WebSocket 的封装，传递给用户
 */
class WebSocket extends events.EventEmitter {
    ws;
    constructor(ws, eventCtx) {
        super();
        this.ws = ws;
        this.ws.onopen = (openEvent) => {
            this.emit('open', {
                ctx: eventCtx,
                type: openEvent.type
            });
        };
        this.ws.onerror = (errorEvent) => {
            this.emit('error', {
                ctx: eventCtx,
                type: errorEvent.type,
                message: errorEvent.message,
                error: errorEvent.error
            });
        };
        this.ws.onclose = (closeEvent) => {
            this.emit('close', {
                ctx: eventCtx,
                type: closeEvent.type,
                code: closeEvent.code,
                reason: closeEvent.reason,
                wasClean: closeEvent.wasClean
            });
        };
        this.ws.onmessage = (messageEvent) => {
            this.emit('message', {
                ctx: eventCtx,
                type: messageEvent.type,
                data: messageEvent.data
            });
        };
        // 隐藏属性
        for (const key of ['ws']) {
            Object.defineProperty(this, key, {
                configurable: false,
                enumerable: false,
                writable: false
            });
        }
    }
    send(buf) {
        this.ws.send(buf);
    }
    close(code, reason) {
        this.ws.close(code, reason);
    }
}
