// @cloudbase/functions-typings/src/cloudbase.ts

export interface TcbExtendedContext {
  envId: string               // 环境 ID
  uin: string                 // 请求的 UIN
  source: string              // 请求来源，如 wx
  serviceName: string         // 服务名称
  serviceVersion: string      // 服务版本
  authMethod?: string         // 认证方式 UNAUTHORIZED | CAM_TC3 ｜ TCB_OAUTH2_B ｜ TCB_OAUTH2_C ｜ WX_SERVER_AUTH
  userType?: string           // 用户类型 NONE | B_SIDE_USER | C_SIDE_USER
  isAdministrator?: boolean   // C 端用户（C_SIDE_USER）是否为管理员
  accessToken?: string        // 调用请求时的 AccessToken
  serviceAccessToken?: string // 服务端注入的 AccessToken，具有管理员权限
  userId?: string             // 请求的用户 ID
  tmpSecret?: {               // 临时凭证
    secretId: string          // secretId
    secretKey: string         // secretKey
    token: string             // token
  }
  wechatContext?: {           // 微信上下文信息
    callId: string            // 微信调用 ID
    source: string            // 请求来源，如 wx
    appId: string             // 访问小程序的 AppID
    openId: string            // 访问用户的 OpenID
    unionId: string           // 访问用户的 UnionID
    fromOpenId?: string       // 环境资源共享时的 fromOpenID
    fromUnionId?: string      // 环境资源共享时的 fromUnionID
    fromAppId?: string        // 环境资源共享时的 fromAppID
  }
}
