// @cloudbase/functions-framework/build/middlewares/accesslog.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTextTypeForTesting = exports.accessLogMiddleware = void 0;
exports.inspectStylize = inspectStylize;
const os_1 = require("os");
const util = __importStar(require("util"));
const whatwg_mimetype_1 = __importDefault(require("whatwg-mimetype"));
const async_context_1 = require("../async-context");
const constants_1 = require("../constants");
const error_1 = require("../error");
const logger_1 = require("../logger");
var LogType;
(function (LogType) {
    LogType["ACCESSLOG"] = "accesslog";
    LogType["USERCODELOG"] = "usercodelog";
})(LogType || (LogType = {}));
const accessLogMiddleware = (option) => {
    const fixedBaseInfo = getFixedBaseLogInfo();
    return async (ctx, next) => {
        const start = new Date();
        await next();
        const end = new Date();
        const { logHeaderBody, logEventContext, logContentEnabled } = option;
        const extendedContext = ctx.extendedContext;
        const errorResult = ctx.state.errorResult;
        const eventID = (0, async_context_1.getCurrentEventID)();
        const reqType = ctx.headers['content-type'] || '';
        const resType = ctx.response.type;
        const log = {
            '@timestamp': start.toISOString(),
            module: '@cloudbase/functions-framework',
            name: '',
            level: 'info',
            startTime: start.toISOString(),
            startAt: start.toISOString(),
            endAt: end.toISOString(),
            logType: LogType.ACCESSLOG,
            ...fixedBaseInfo,
            eventId: eventID,
            requestId: eventID,
            method: ctx.method,
            url: ctx.href,
            path: ctx.path,
            status: ctx.status,
            httphost: ctx.headers.host,
            connection: ctx.headers.connection,
            upgrade: ctx.headers.upgrade,
            referrer: ctx.headers.referer,
            origin: ctx.headers.origin,
            forwarded: ctx.headers.forwarded,
            xff: ctx.get('X-Forwarded-For'),
            ua: ctx.get('user-agent'),
            uasecch: ctx.get('sec-ch-ua'),
            sourceIp: ctx.get('x-real-ip'),
            reqContentType: reqType,
            reqSize: ctx.request.length,
            resContentType: resType,
            resSize: ctx.response.length,
            timeCost: end.valueOf() - start.valueOf(),
            userCodeTimeCost: ctx.state.userCodeTimeCost,
            code: errorResult?.code,
            message: errorResult?.message,
            userId: extendedContext?.userId,
            source: extendedContext?.source,
            matchedRoute: ctx.state.matchedRoute?.path,
            functionName: ctx.state.matchedRoute?.functionName
        };
        if (logHeaderBody) {
            // 因为 context&event 包含大部分信息，这里默认不打印比较好
            log.reqHeader = inspect(ctx.httpContext?.headers);
            log.resHeader = inspect({ ...ctx.response.headers });
            log.reqBody = isTextType(reqType) ? inspect(ctx.request.body) : '';
            log.resBody = isTextType(resType) ? inspect(ctx.response.body) : '';
        }
        if (ctx.method === 'OPTIONS' && logHeaderBody === false) {
            // 如果是 OPTIONS 请求打印请求头相应头
            log.reqHeader = inspect(ctx.httpContext?.headers);
            log.resHeader = inspect({ ...ctx.response.headers });
        }
        if (logEventContext) {
            log.event = inspect(ctx.state.event);
            log.context = inspect(ctx.state.contextInjected);
        }
        if (ctx.state.codeInfo) {
            log.name = ctx.state.codeInfo.name || '';
            log.userCodeVersion = `${ctx.state.codeInfo.name}/${ctx.state.codeInfo.version}`;
        }
        if (ctx.state.error) {
            const error = ctx.state.error;
            log.error = error.message;
            log.stack = (0, error_1.decorateErrorStack)(error);
            log.stackFull = error?.stack;
        }
        if (logContentEnabled) {
            /* eslint-disable max-len */
            log['__CONTENT__'] = `method:${log.method} path:${log.path} status:${log.status} timeCost:${log.timeCost}ms bytes:${log.reqSize || 0}b/${log.resSize || 0}b userid:${log.userId || ''} ip:${log.sourceIp || ''} code:${log.code || ''}`;
            /* eslint-enable max-len */
        }
        logger_1.functionsLogger.logAccesslog('info', JSON.stringify(log));
        logger_1.functionsLogger.logUserCodelog();
    };
};
exports.accessLogMiddleware = accessLogMiddleware;
function inspect(obj) {
    /* eslint-disable @typescript-eslint/no-explicit-any */
    return util.inspect(obj, {
        depth: 5,
        maxStringLength: 1024,
        maxArrayLength: 100,
        stylize: inspectStylize(200)
    });
    /* eslint-enable @typescript-eslint/no-explicit-any */
}
function inspectStylize(maxLen) {
    return (str, type) => {
        // 对象的 Key
        if (type === 'name') {
            return str;
        }
        // 对象的值，type 可能存在多种类型
        if (type === 'string') {
            if (str.length > maxLen) {
                return '\'' + str.slice(1, maxLen) + ' ... ' + `${str.length - maxLen}(/${str.length}) more characters'`;
            }
            return str;
        }
        else if (type === 'special') {
            // 函数
            return `'${str}'`;
        }
        return str;
    };
}
// https://mimesniff.spec.whatwg.org/#mime-type-groups
const isTextType = (declaration) => {
    const type = whatwg_mimetype_1.default.parse(declaration);
    if (!type) {
        return false;
    }
    if (type.isHTML() || type.isXML() || type.isJavaScript()) {
        return true;
    }
    if (type.subtype.endsWith('+json') || ['application/json', 'text/json'].includes(type.essence)) {
        return true;
    }
    if (type.type === 'text') {
        return true;
    }
    if (type.subtype === 'x-www-form-urlencoded') {
        return true;
    }
    return false;
};
const getFixedBaseLogInfo = () => {
    return {
        hostname: (0, os_1.hostname)(),
        pid: process.pid,
        version: constants_1.SDKVersion
    };
};
/**
 * This should only be accessed in test files
 */
exports.isTextTypeForTesting = isTextType;
