// @cloudbase/functions-framework/build/middlewares/cors.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.allowOriginForTesting = exports.corsMiddleware = void 0;
const cors_1 = __importDefault(require("@koa/cors"));
const defaultAllowedLocalOrigin = ['localhost', '127.0.0.1'];
const allowOrigin = (origin, allowed) => {
    try {
        // 浏览器在 file:/// 页面发起请求时 origin 为 null
        // 这里构造构造成 url 格式，避免 URL 解析失败
        if (origin === 'null') {
            origin = 'protocol://null';
        }
        // 解析 origin，如果解析失败直接返回 false
        const url = new URL(origin);
        return allowed.some((pattern) => {
            if (pattern === '*') {
                return true;
            }
            // 支持通配符匹配 (*.example.com)
            if (pattern.startsWith('*.')) {
                const domain = pattern.slice(2); // 去掉 *.
                return url.hostname.endsWith(`.${domain}`);
            }
            // 精确匹配主机名
            return url.hostname === pattern;
        });
    }
    catch (error) {
        // origin 不是合法的 URL，返回 false
        return false;
    }
};
const corsMiddleware = (enable, allowedOrigins) => {
    const allowedOrigin = allowedOrigins && allowedOrigins.length > 0
        ? defaultAllowedLocalOrigin.concat(allowedOrigins)
        : defaultAllowedLocalOrigin;
    return async (ctx, next) => {
        if (!enable) {
            return await next();
        }
        // tcb-ff.js -w --enableCors=true --allowedOrigins="*" --allowedAccounts=abc:123,efg:456
        // curl -v -X OPTIONS -H "Access-Control-Request-Method: GET" -H "Origin: http://127.0.0.1:80" http://127.0.0.1:3000
        return await (0, cors_1.default)({
            origin: (ctx) => {
                if (!ctx.headers.origin) {
                    return '';
                }
                if (allowOrigin(ctx.headers.origin, allowedOrigin)) {
                    return ctx.headers.origin;
                }
                return '';
            },
            allowMethods: '*',
            allowHeaders: '*',
            exposeHeaders: '*',
            maxAge: 10, // 标准规定默认值为 5s
            credentials: true,
            keepHeadersOnError: false
        })(ctx, next);
    };
};
exports.corsMiddleware = corsMiddleware;
/**
 * This should only be accessed in test files
 */
exports.allowOriginForTesting = allowOrigin;
